File: morphac/math/geometry/test/shapes_test.cc
#include "math/geometry/include/shapes.h"

#include "gtest/gtest.h"

namespace {

using Eigen::Vector2d;

using morphac::common::aliases::Point;
using morphac::math::geometry::ArcShape;
using morphac::math::geometry::CircleShape;
using morphac::math::geometry::RectangleShape;
using morphac::math::geometry::RoundedRectangleShape;
using morphac::math::geometry::TriangleShape;
using morphac::utils::IsEqual;

class ShapesTest : public ::testing::Test {
 protected:
  ShapesTest() {
    // Set random seed for Eigen.
    srand(7);
  }

  void SetUp() override {}

  // Standard shapes
  ArcShape arc_shape_{M_PI / 4, M_PI, 2., Point(-1., 2.)};
  CircleShape circle_shape_{2., Point(-1., 2.)};
  RectangleShape rectangle_shape_{3., 1.5, M_PI / 4, Point(-1., 2.)};
  RoundedRectangleShape rounded_rectangle_shape_{3., 1.5, M_PI / 4, 0.5,
                                                 Point(-1., 2.)};
  TriangleShape triangle_shape_{3., 1.5, M_PI / 4., Point(-1., 2.)};
};

TEST_F(ShapesTest, ArcShape) {
  ASSERT_TRUE(arc_shape_.center.isApprox(Vector2d(-1, 2)));
  ASSERT_TRUE(IsEqual(arc_shape_.start_angle, M_PI / 4));
  ASSERT_TRUE(IsEqual(arc_shape_.end_angle, M_PI));
  ASSERT_TRUE(IsEqual(arc_shape_.radius, 2.));
}

TEST_F(ShapesTest, DefaultCenterArcShape) {
  ASSERT_TRUE(ArcShape(0., M_PI / 2, 2.).center.isApprox(Vector2d::Zero()));
}

TEST_F(ShapesTest, InvalidArcShape) {
  // Invalid radius.
  ASSERT_THROW(ArcShape(0., M_PI / 2, -0.1), std::invalid_argument);

  // Invalid start and end angles (equal).
  ASSERT_THROW(ArcShape(M_PI / 2., M_PI / 2, 1.), std::invalid_argument);
}

TEST_F(ShapesTest, CircleShape) {
  ASSERT_TRUE(circle_shape_.center.isApprox(Vector2d(-1, 2)));
  ASSERT_TRUE(IsEqual(circle_shape_.radius, 2.));
}

TEST_F(ShapesTest, DefaultCenterCircleShape) {
  ASSERT_TRUE(CircleShape(2.).center.isApprox(Vector2d::Zero()));
}

TEST_F(ShapesTest, InvalidCircleShape) {
  // Invalid radius.
  ASSERT_THROW(CircleShape(-0.1), std::invalid_argument);
}

TEST_F(ShapesTest, RectangleShape) {
  ASSERT_TRUE(rectangle_shape_.center.isApprox(Vector2d(-1, 2)));
  ASSERT_TRUE(IsEqual(rectangle_shape_.size_x, 3.));
  ASSERT_TRUE(IsEqual(rectangle_shape_.size_y, 1.5));
  ASSERT_TRUE(IsEqual(rectangle_shape_.angle, M_PI / 4));
}

TEST_F(ShapesTest, DefaultCenterRectangleShape) {
  ASSERT_TRUE(RectangleShape(1., 1., 0.).center.isApprox(Vector2d::Zero()));
}

TEST_F(ShapesTest, InvalidRectangleShape) {
  // Invalid size_x.
  ASSERT_THROW(RectangleShape(-0.1, 1., 0.), std::invalid_argument);
  // Invalid size_y.
  ASSERT_THROW(RectangleShape(1., -0.1, 0.), std::invalid_argument);
}

TEST_F(ShapesTest, RoundedRectangleShape) {
  ASSERT_TRUE(rounded_rectangle_shape_.center.isApprox(Vector2d(-1, 2)));
  ASSERT_TRUE(IsEqual(rounded_rectangle_shape_.size_x, 3.));
  ASSERT_TRUE(IsEqual(rounded_rectangle_shape_.size_y, 1.5));
  ASSERT_TRUE(IsEqual(rounded_rectangle_shape_.angle, M_PI / 4));
  ASSERT_TRUE(IsEqual(rounded_rectangle_shape_.radius, 0.5));
}

TEST_F(ShapesTest, DefaultCenterRoundedRectangleShape) {
  ASSERT_TRUE(
      RoundedRectangleShape(1., 1., 0., 0.1).center.isApprox(Vector2d::Zero()));
}

TEST_F(ShapesTest, InvalidRoundedRectangleShape) {
  // Invalid size_x.
  ASSERT_THROW(RoundedRectangleShape(-0.1, 1., 0., 0.1), std::invalid_argument);
  // Invalid size_y.
  ASSERT_THROW(RoundedRectangleShape(1., -0.1, 0., 0.1), std::invalid_argument);
  // Invalid radius.
  ASSERT_THROW(RoundedRectangleShape(1., 1., 0., -0.1), std::invalid_argument);
  // The radius must also be smaller than half of the sizes.
  ASSERT_THROW(RoundedRectangleShape(3., 2., 0., 1.1), std::invalid_argument);
  ASSERT_THROW(RoundedRectangleShape(2., 3., 0., 1.1), std::invalid_argument);
}

TEST_F(ShapesTest, TriangleShape) {
  ASSERT_TRUE(triangle_shape_.center.isApprox(Vector2d(-1, 2)));
  ASSERT_TRUE(IsEqual(triangle_shape_.base, 3.));
  ASSERT_TRUE(IsEqual(triangle_shape_.height, 1.5));
  ASSERT_TRUE(IsEqual(triangle_shape_.angle, M_PI / 4));
}

TEST_F(ShapesTest, DefaultCenterTriangleShape) {
  ASSERT_TRUE(TriangleShape(1., 1., 0.).center.isApprox(Vector2d::Zero()));
}

TEST_F(ShapesTest, InvalidTriangleShape) {
  // Invalid base.
  ASSERT_THROW(TriangleShape(-0.1, 1., 0.), std::invalid_argument);
  // Invalid height.
  ASSERT_THROW(TriangleShape(1., -0.1, 0.), std::invalid_argument);
}

}  // namespace

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/math/geometry/python/test/shapes_test.py
import pytest

import numpy as np

from morphac.math.geometry import (
    ArcShape,
    CircleShape,
    RectangleShape,
    RoundedRectangleShape,
    TriangleShape,
)


@pytest.fixture()
def generate_shapes():

    arc = ArcShape(
        center=[-1.0, 2.0], start_angle=np.pi / 4, end_angle=np.pi / 2, radius=2.0
    )

    circle = CircleShape(center=[-1.0, 2.0], radius=2.0)

    rectangle = RectangleShape(
        center=[-1.0, 2.0], size_x=3.0, size_y=1.5, angle=np.pi / 4
    )

    rounded_rectangle = RoundedRectangleShape(
        center=[-1.0, 2.0], size_x=3.0, size_y=1.5, angle=np.pi / 4, radius=0.5
    )

    triangle = TriangleShape(center=[-1.0, 2.0], base=3.0, height=1.5, angle=np.pi / 4)

    return arc, circle, rectangle, rounded_rectangle, triangle


def test_arc_shape(generate_shapes):
    arc, _, _, _, _ = generate_shapes

    assert np.isclose(arc.start_angle, np.pi / 4)
    assert np.isclose(arc.end_angle, np.pi / 2)
    assert np.isclose(arc.radius, 2.0)
    assert np.allclose(arc.center, [-1, 2])


def test_invalid_arc_shape():
    # Invalid radius.
    with pytest.raises(ValueError):
        _ = ArcShape([0, 0], 0.0, np.pi / 2, -0.1)
    # Invalid start and end angles (equal).
    with pytest.raises(ValueError):
        _ = ArcShape([0, 0], np.pi / 2, np.pi / 2, 1.0)


def test_circle_shape(generate_shapes):
    _, circle, _, _, _ = generate_shapes

    assert np.isclose(circle.radius, 2.0)
    assert np.allclose(circle.center, [-1, 2])


def test_invalid_circle_shape():
    # Invalid radius.
    with pytest.raises(ValueError):
        _ = CircleShape([0, 0], -0.1)


def test_rectangle_shape(generate_shapes):
    _, _, rectangle, _, _ = generate_shapes

    assert np.isclose(rectangle.size_x, 3)
    assert np.isclose(rectangle.size_y, 1.5)
    assert np.isclose(rectangle.angle, np.pi / 4)
    assert np.allclose(rectangle.center, [-1, 2])


def test_invalid_rectangle_shape():
    # Invalid size_x.
    with pytest.raises(ValueError):
        _ = RectangleShape([0, 0], -0.1, 1.0, 0.0)
    # Invalid size_y.
    with pytest.raises(ValueError):
        _ = RectangleShape([0, 0], 1.0, -0.1, 0.0)


def test_rounded_rectangle_shape(generate_shapes):
    _, _, _, rounded_rectangle, _ = generate_shapes

    assert np.isclose(rounded_rectangle.size_x, 3)
    assert np.isclose(rounded_rectangle.size_y, 1.5)
    assert np.isclose(rounded_rectangle.angle, np.pi / 4)
    assert np.isclose(rounded_rectangle.radius, 0.5)
    assert np.allclose(rounded_rectangle.center, [-1, 2])


def test_invalid_rounded_rectangle_shape():
    # Invalid size_x.
    with pytest.raises(ValueError):
        _ = RoundedRectangleShape([0, 0], -0.1, 1.0, 0.0, 0.1)
    # Invalid size_y.
    with pytest.raises(ValueError):
        _ = RoundedRectangleShape([0, 0], 1.0, -0.1, 0.0, 0.1)
    # Invalid radius.
    with pytest.raises(ValueError):
        _ = RoundedRectangleShape([0, 0], 1.0, 1.0, 0.0, -0.1)
    # The radius must also be smaller than half of the sizes.
    with pytest.raises(ValueError):
        _ = RoundedRectangleShape([0, 0], 3.0, 2.0, 0.0, 1.1)
    with pytest.raises(ValueError):
        _ = RoundedRectangleShape([0, 0], 2.0, 3.0, 0.0, 1.1)


def test_triangle_shape(generate_shapes):
    _, _, _, _, triangle_shape = generate_shapes

    assert np.isclose(triangle_shape.base, 3)
    assert np.isclose(triangle_shape.height, 1.5)
    assert np.isclose(triangle_shape.angle, np.pi / 4)
    assert np.allclose(triangle_shape.center, [-1, 2])


def test_invalid_triangle_shape():
    # Invalid base.
    with pytest.raises(ValueError):
        _ = TriangleShape([0, 0], -0.1, 1.0, 0.0)
    # Invalid height.
    with pytest.raises(ValueError):
        _ = TriangleShape([0, 0], 1.0, -0.1, 0.0)
